File: backend/app/services/plan_catalog.py
"""Каталог тарифов ModelMate: tier × billing_plan × период."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

from app.services.billing_plan import (
    BILLING_PLAN_CREDITS,
    BILLING_PLAN_PRO,
    BILLING_PLAN_STANDARD,
    normalize_billing_plan,
)
from app.services.referral import referral_public_dict

PlanTier = Literal["solo", "pro", "studio"]
BillingPeriod = Literal["month", "year"]
BillingPlanKind = Literal["standard", "pro"]

TIER_SOLO: PlanTier = "solo"
TIER_PRO: PlanTier = "pro"
TIER_STUDIO: PlanTier = "studio"
PERIOD_MONTH: BillingPeriod = "month"
PERIOD_YEAR: BillingPeriod = "year"

TIERS: tuple[PlanTier, ...] = (TIER_SOLO, TIER_PRO, TIER_STUDIO)
PERIODS: tuple[BillingPeriod, ...] = (PERIOD_MONTH, PERIOD_YEAR)

# product id (старые и новые) → актуальный тариф
LEGACY_PRODUCT_MAP: dict[str, str] = {
    "sub_managed_month": "sub_standard_solo_month",
    "sub_byok_month": "sub_pro_solo_month",
    "sub_managed_solo_month": "sub_standard_solo_month",
    "sub_managed_pro_month": "sub_standard_pro_month",
    "sub_managed_studio_month": "sub_standard_studio_month",
    "sub_managed_solo_year": "sub_standard_solo_year",
    "sub_managed_pro_year": "sub_standard_pro_year",
    "sub_managed_studio_year": "sub_standard_studio_year",
    "sub_byok_solo_month": "sub_pro_solo_month",
    "sub_byok_pro_month": "sub_pro_pro_month",
    "sub_byok_studio_month": "sub_pro_studio_month",
    "sub_byok_solo_year": "sub_pro_solo_year",
    "sub_byok_pro_year": "sub_pro_pro_year",
    "sub_byok_studio_year": "sub_pro_studio_year",
}


@dataclass(frozen=True, slots=True)
class PlanLimits:
    max_users: int
    max_models: int
    max_dialogs_per_month: int | None
    max_grok_per_month: int | None


@dataclass(frozen=True, slots=True)
class PlanSpec:
    product: str
    billing_plan: BillingPlanKind
    tier: PlanTier
    period: BillingPeriod
    price_rub: int
    title_ru: str
    managed_monthly_credits: int
    limits: PlanLimits
    popular: bool = False


CREDITS_PLAN_LIMITS = PlanLimits(
    max_users=1,
    max_models=1,
    max_dialogs_per_month=0,
    max_grok_per_month=None,
)


def _limits(tier: PlanTier) -> PlanLimits:
    if tier == TIER_SOLO:
        return PlanLimits(
            max_users=1,
            max_models=1,
            max_dialogs_per_month=1000,
            max_grok_per_month=500,
        )
    if tier == TIER_PRO:
        return PlanLimits(
            max_users=3,
            max_models=3,
            max_dialogs_per_month=5000,
            max_grok_per_month=2000,
        )
    return PlanLimits(
        max_users=10,
        max_models=10,
        max_dialogs_per_month=None,
        max_grok_per_month=10000,
    )


def _product(billing: BillingPlanKind, tier: PlanTier, period: BillingPeriod) -> str:
    return f"sub_{billing}_{tier}_{period}"


def managed_period_credits(spec: PlanSpec) -> int:
    """Кредиты при оплате Standard за период."""
    if spec.billing_plan != BILLING_PLAN_STANDARD or spec.managed_monthly_credits <= 0:
        return 0
    if spec.period == PERIOD_YEAR:
        return spec.managed_monthly_credits * 12
    return spec.managed_monthly_credits


def _build_specs() -> dict[str, PlanSpec]:
    prices_month_pro: dict[PlanTier, int] = {
        TIER_SOLO: 990,
        TIER_PRO: 2490,
        TIER_STUDIO: 5990,
    }
    prices_month_standard: dict[PlanTier, int] = {
        TIER_SOLO: 1990,
        TIER_PRO: 4990,
        TIER_STUDIO: 11990,
    }
    prices_year_pro: dict[PlanTier, int] = {
        TIER_SOLO: 8900,
        TIER_PRO: 22400,
        TIER_STUDIO: 53900,
    }
    prices_year_standard: dict[PlanTier, int] = {
        TIER_SOLO: 17900,
        TIER_PRO: 44900,
        TIER_STUDIO: 107900,
    }
    standard_credits: dict[PlanTier, int] = {
        TIER_SOLO: 150,
        TIER_PRO: 400,
        TIER_STUDIO: 1200,
    }
    titles: dict[tuple[BillingPlanKind, PlanTier], str] = {
        (BILLING_PLAN_PRO, TIER_SOLO): "Pro Solo",
        (BILLING_PLAN_PRO, TIER_PRO): "Pro Pro",
        (BILLING_PLAN_PRO, TIER_STUDIO): "Pro Studio",
        (BILLING_PLAN_STANDARD, TIER_SOLO): "Standard Solo",
        (BILLING_PLAN_STANDARD, TIER_PRO): "Standard Pro",
        (BILLING_PLAN_STANDARD, TIER_STUDIO): "Standard Studio",
    }
    out: dict[str, PlanSpec] = {}
    for billing in (BILLING_PLAN_PRO, BILLING_PLAN_STANDARD):
        month_prices = prices_month_pro if billing == BILLING_PLAN_PRO else prices_month_standard
        year_prices = prices_year_pro if billing == BILLING_PLAN_PRO else prices_year_standard
        for tier in TIERS:
            for period, price_map in ((PERIOD_MONTH, month_prices), (PERIOD_YEAR, year_prices)):
                prod = _product(billing, tier, period)
                out[prod] = PlanSpec(
                    product=prod,
                    billing_plan=billing,
                    tier=tier,
                    period=period,
                    price_rub=price_map[tier],
                    title_ru=titles[(billing, tier)],
                    managed_monthly_credits=(
                        standard_credits[tier] if billing == BILLING_PLAN_STANDARD else 0
                    ),
                    limits=_limits(tier),
                    popular=(tier == TIER_PRO),
                )
    return out


PLAN_SPECS: dict[str, PlanSpec] = _build_specs()


def normalize_plan_tier(raw: str | None) -> PlanTier:
    s = (raw or TIER_SOLO).strip().lower()
    if s in TIERS:
        return s  # type: ignore[return-value]
    return TIER_SOLO


def resolve_product_id(product: str) -> str:
    p = (product or "").strip()
    return LEGACY_PRODUCT_MAP.get(p, p)


def get_plan_spec(product: str) -> PlanSpec | None:
    return PLAN_SPECS.get(resolve_product_id(product))


def list_subscription_products() -> list[PlanSpec]:
    return sorted(PLAN_SPECS.values(), key=lambda s: (s.billing_plan, TIERS.index(s.tier), s.period))


def tier_label_ru(tier: PlanTier) -> str:
    return {"solo": "Solo", "pro": "Pro", "studio": "Studio"}[tier]


def plan_display_name(billing_plan: str | None, plan_tier: str | None) -> str:
    bp = normalize_billing_plan(billing_plan)
    if bp == BILLING_PLAN_CREDITS:
        return "Credits"
    tier = normalize_plan_tier(plan_tier)
    mode = "Pro" if bp == BILLING_PLAN_PRO else "Standard"
    return f"{mode} {tier_label_ru(tier)}"


def catalog_public_dict() -> dict:
    """Сериализация для /api/health и маркетинга."""
    plans = []
    for spec in list_subscription_products():
        lim = spec.limits
        plans.append(
            {
                "product": spec.product,
                "billing_plan": spec.billing_plan,
                "tier": spec.tier,
                "period": spec.period,
                "price_rub": spec.price_rub,
                "title": spec.title_ru,
                "popular": spec.popular,
                "managed_monthly_credits": spec.managed_monthly_credits,
                "subscription_monthly_credits": spec.managed_monthly_credits,
                "managed_period_credits": managed_period_credits(spec),
                "subscription_period_credits": managed_period_credits(spec),
                "limits": {
                    "max_users": lim.max_users,
                    "max_models": lim.max_models,
                    "max_dialogs_per_month": lim.max_dialogs_per_month,
                    "max_grok_per_month": lim.max_grok_per_month,
                },
            }
        )
    return {
        "plans": plans,
        "legacy_products": LEGACY_PRODUCT_MAP,
        "billing_plan_kinds": [BILLING_PLAN_CREDITS, BILLING_PLAN_STANDARD, BILLING_PLAN_PRO],
        "wavespeed_referral_url": "https://wavespeed.ai/?ref=modelmate",
        "referral": referral_public_dict(),
    }

File: frontend/src/billing/planCatalog.ts
/** Каталог тарифов (синхрон с backend plan_catalog.py). */

export const WAVESPEED_REF_URL = 'https://wavespeed.ai/?ref=modelmate'

export type PlanTier = 'solo' | 'pro' | 'studio'
export type BillingPlanKind = 'standard' | 'pro'
export type CreditsPlanKind = 'credits' | 'standard' | 'pro'
export type BillingPeriod = 'month' | 'year'

export interface PlanLimits {
  max_users: number
  max_models: number
  max_dialogs_per_month: number | null
  max_grok_per_month: number | null
}

export interface CatalogPlan {
  product: string
  billing_plan: BillingPlanKind
  tier: PlanTier
  period: BillingPeriod
  price_rub: number
  title: string
  popular: boolean
  managed_monthly_credits: number
  /** Кредиты за оплаченный период (год = 12× monthly). С бэкенда; иначе считается на клиенте. */
  managed_period_credits?: number
  limits: PlanLimits
}

const LEGACY_BILLING_ALIASES: Record<string, CreditsPlanKind> = {
  managed: 'standard',
  byok: 'pro',
}

/** Нормализация billing_plan с учётом legacy managed/byok. */
export function normalizeBillingPlan(raw: string | null | undefined): CreditsPlanKind {
  const s = (raw || 'standard').trim().toLowerCase()
  if (s in LEGACY_BILLING_ALIASES) return LEGACY_BILLING_ALIASES[s]
  if (s === 'credits' || s === 'standard' || s === 'pro') return s
  return 'standard'
}

/** Кредиты при оплате подписки Standard за выбранный период. */
export function managedPeriodCredits(
  plan: Pick<CatalogPlan, 'billing_plan' | 'period' | 'managed_monthly_credits' | 'managed_period_credits'>,
): number {
  if (plan.managed_period_credits != null && plan.managed_period_credits > 0) {
    return plan.managed_period_credits
  }
  if (plan.billing_plan !== 'standard' || plan.managed_monthly_credits <= 0) return 0
  return plan.period === 'year' ? plan.managed_monthly_credits * 12 : plan.managed_monthly_credits
}

export const LIMITS: Record<PlanTier, PlanLimits> = {
  solo: { max_users: 1, max_models: 1, max_dialogs_per_month: 1000, max_grok_per_month: 500 },
  pro: { max_users: 3, max_models: 3, max_dialogs_per_month: 5000, max_grok_per_month: 2000 },
  studio: { max_users: 10, max_models: 10, max_dialogs_per_month: null, max_grok_per_month: 10000 },
}

function spec(
  billing: BillingPlanKind,
  tier: PlanTier,
  period: BillingPeriod,
  price: number,
  credits: number,
): CatalogPlan {
  const mode = billing === 'pro' ? 'Pro' : 'Standard'
  const tierLabel = { solo: 'Solo', pro: 'Pro', studio: 'Studio' }[tier]
  return {
    product: `sub_${billing}_${tier}_${period}`,
    billing_plan: billing,
    tier,
    period,
    price_rub: price,
    title: `${mode} ${tierLabel}`,
    popular: tier === 'pro',
    managed_monthly_credits: credits,
    limits: LIMITS[tier],
  }
}

/** Fallback, если /api/health недоступен. */
export const FALLBACK_CATALOG_PLANS: CatalogPlan[] = [
  spec('pro', 'solo', 'month', 990, 0),
  spec('pro', 'pro', 'month', 2490, 0),
  spec('pro', 'studio', 'month', 5990, 0),
  spec('standard', 'solo', 'month', 1990, 150),
  spec('standard', 'pro', 'month', 4990, 400),
  spec('standard', 'studio', 'month', 11990, 1200),
  spec('pro', 'solo', 'year', 8900, 0),
  spec('pro', 'pro', 'year', 22400, 0),
  spec('pro', 'studio', 'year', 53900, 0),
  spec('standard', 'solo', 'year', 17900, 150),
  spec('standard', 'pro', 'year', 44900, 400),
  spec('standard', 'studio', 'year', 107900, 1200),
]

function normalizeCatalogPlan(raw: unknown): CatalogPlan | null {
  if (!raw || typeof raw !== 'object') return null
  const p = raw as Record<string, unknown>
  const billing = normalizeBillingPlan(String(p.billing_plan ?? ''))
  if (billing !== 'standard' && billing !== 'pro') return null
  return { ...(raw as CatalogPlan), billing_plan: billing }
}

export function parseCatalogFromHealth(
  health: { billing_catalog?: { plans?: unknown } } | null | undefined,
): CatalogPlan[] {
  const raw = health?.billing_catalog?.plans
  if (Array.isArray(raw) && raw.length) {
    const plans = raw.map(normalizeCatalogPlan).filter((p): p is CatalogPlan => p != null)
    if (plans.length) return plans
  }
  return FALLBACK_CATALOG_PLANS
}

export function filterPlans(
  plans: CatalogPlan[],
  billing: BillingPlanKind,
  period: BillingPeriod,
): CatalogPlan[] {
  return plans.filter((p) => p.billing_plan === billing && p.period === period)
}

export function tierFeatures(
  tier: PlanTier,
  billing: BillingPlanKind,
  period: BillingPeriod = 'month',
  managedMonthlyCredits?: number,
): string[] {
  const l = LIMITS[tier]
  const users =
    l.max_users === 1
      ? '1 пользователь'
      : `До ${l.max_users} пользователей`
  const models =
    l.max_models === 1 ? '1 модель' : `До ${l.max_models} моделей`
  const base = [
    users,
    models,
    billing === 'pro'
      ? 'Свой ключ WaveSpeed — без списаний на платформе'
      : 'Кредиты на генерацию включены',
    'Чат Fanvue + Telegram',
    'Авто-перевод переписок',
    'История генераций',
  ]
  if (billing === 'standard' && LIMITS[tier]) {
    const monthly = managedMonthlyCredits ?? { solo: 150, pro: 400, studio: 1200 }[tier]
    const total = period === 'year' ? monthly * 12 : monthly
    base[2] =
      period === 'year'
        ? `${total.toLocaleString('ru-RU')} кредитов за год (${monthly} / мес.)`
        : `${total} кредитов за месяц`
  }
  if (tier === 'pro') base.push('Командная работа и роли')
  if (tier === 'studio') base.push('Расширенные лимиты диалогов')
  return base
}
